Add option to skip installing directory dependencies

adriangb opened the pull request:
This serves the same purpose as `--only-root` but for path deps (which just like the "root" will bust caches if any source file changes).

If/once https://github.com/python-poetry/poetry-core/pull/520 gets merged this will allow writing a Dockerfile along the lines of:

```
FROM python

COPY pyproject.toml poetry.lock .
RUN pip install poetry && poetry install --no-root --no-path
COPY src/ ./src
COPY folder-of-path-deps/ folder-of-path-deps/
RUN poetry install
```

I chose `--no-path` because of it's parallels to `--no-root`, but I'm open to other options.

Closes #6845
Closes #6850 
Closes #3671

adriangb commented on docs/faq.md:
````diff
@@ -152,3 +152,37 @@ This is done so to be compliant with the broader Python ecosystem.
 [PEP 440](https://peps.python.org/pep-0440), third party tools will be unable to parse the version correctly.
+
+
+### Poetry busts my Docker cache because it requires me to COPY my source files in before installing 3rd party dependencies
+
+By default running `poetry install ...` requires you to have your source files present (both the "root" package and any `path` dependencies you might have).
+This interacts poorly with Docker's caching mechanisms because any change to the source file will make any layers (subsequent commands in your Dockerfile) re-run.
+For example, you might have a Dockerfile that looks something like this:
+
+```text
+FROM python
+COPY pyproject.toml poetry.lock .
+COPY src/ ./src
+RUN pip install poetry && poetry install --no-dev
+```
+
+The `RUN` instruction will always re-run, which forces all 3rd party dependencies (likely the slowest step out of these) to re-run if you changed any files in `src/`.
+
+To avoid this cache busting you can split this into two steps:
+
+1. Install 3rd party dependencies.
+2. Copy over your source code and install just the source code.
+
+This might look something like this:
+
+```text
+FROM python
+COPY pyproject.toml poetry.lock .
+RUN pip install poetry && poetry install --no-root --no-path
+COPY src/ ./src
+RUN poetry install --no-dev
+```
+
+The two key options we are using here are `--no-root` (skips installing the project source) and `--no-path` (skips installing any local path dependencies, you can skip this if you don't have any).
+[More information on the options available for `poetry install`]({{< relref "cli#install" >}}).
````
I'm just kinda throwing this out here. I know we probably need more documentation on Docker best practices in general but I think that's way beyond the scope of this PR. Even this little bit is maybe pushing it, but I'm hoping that it's okay to put in here if I restrict the scope to just the impact of `--no-root` and `--no-path` (as opposed to jumping into multistage builds and whatnot).

adriangb commented on tests/console/commands/test_install.py:
```diff
@@ -195,3 +195,15 @@ def test_dry_run_populates_installer(tester: CommandTester, mocker: MockerFixtur
+
+    tester.execute("--no-path")
+
+    assert tester.command.installer._skip_path is True
```
I don't really like this but it's exactly what all of the other tests do so 🤷🏻‍♂️

adriangb commented on tests/installation/test_installer.py:
```diff
@@ -60,12 +60,12 @@ class Executor(BaseExecutor):
 
-        self._installs: list[DependencyPackage] = []
+        self._installs: list[Package] = []
         self._updates: list[DependencyPackage] = []
         self._uninstalls: list[DependencyPackage] = []
 
     @property
-    def installations(self) -> list[DependencyPackage]:
+    def installations(self) -> list[Package]:
```
These types seem to be wrong so I'm fixing them. I bet it's the same for `_updates` but I wasn't touching them in this PR so I decided to leave em alone.

adriangb commented:
I think this fixes https://github.com/python-poetry/poetry/issues/668

adriangb commented:
@radoering while we wait for review on https://github.com/python-poetry/poetry-core/pull/520, would you mind taking a look at this and giving some thoughts? Thanks!

radoering commented:
If I understand correctly, the main motivation for this feature is the Docker cache. Before taking a look into the implementation details, I'd like to here the opinions of other maintainers if this feature makes sense for them. Especially @neersighted because he has much more experience with Docker than I do.

Just one question: Does it make sense to lump directory and file dependencies together (since both are path dependencies) or should it only consider directory dependencies?

adriangb commented:
> Just one question: Does it make sense to lump directory and file dependencies together (since both are path dependencies) or should it only consider directory dependencies?

Great question. I’m not super familiar with workflows around file dependencies, I’ve never used them. The fundamental question as far as caches are concerned is “does this likely change every build (source code) or only rarely (dependencies)?”.
The main use case I associate file dependencies with is vendoring, in which case they’re not changing any more often than other leaf dependencies. What makes directory dependencies special is that they’re often actually your own source code.
We could always add another flag allowing users to choose depending on their use case. It’s a bit more complexity but avoids painting ourselves into a corner if we miss some use cases.

neersighted commented:
I'm not convinced this is useful, as in the use-case shown here (not building a package, just using Poetry to set up an environment), groups could easily be used. I don't think path dependencies should be special at all; encouraging users to think of them as 'special' takes us down a path we have strictly wanted to avoid (e.g. leaking metadata beyond the Core Metadata spec between layers of the dependency tree because Poetry is used all the way down), and reduces the interoperability of Poetry with the rest of the ecosystem.

I think that documenting the use of this pattern, but with groups, can help the people who are trying to do monorepos today, and that a workspace-like approach like suggested in #2270 is what we should explore, instead of trying to tear down the Core Metadata/PEP 517 wall between Poetry projects.

adriangb commented:
> in the use-case shown here (not building a package, just using Poetry to set up an environment), groups could easily be used

could you explain how? I use dependency groups in conjunction with path dependencies in a real world monorepo and it works great aside from docker caching. I don’t see how “using dependency groups” solves anything?

neersighted commented:
Hmm, I did not realize that this PR was targeted at the pattern in your example repo there. While I brought up the overall direction for better supporting monorepos as kind of a extension of my thinking re: this PR, I did not realize that is at the core of what you want to do.

I think that makes my objection firmer -- special support for using path dependencies in this way (with a superproject that delegates to subprojects) is not desirable because it is ultimately incomplete unless we tear down the metadata barrier, and all the designs along that line have been rejected to date. It's possible we decide that the subproject approach is not tenable and go that way after all, but I think it would have to be comprehensive and not piecemeal. It would probably be better if it was a new dependency type if that were the case, instead of overloading `path =`, but I digress.

Regarding the use case you have, I'm afraid I still don't see how what you want to do is not possible with dependency groups. Could one not do the following (with all your groups set to optional)?

``` dockerfile
FROM python:3.11

COPY pyproject.toml poetry.lock .
RUN pip install poetry && poetry install --no-root
COPY src/ ./src
RUN poetry install --only-root
COPY folder-of-path-deps/ folder-of-path-deps/
RUN poetry install --no-root --with <path-dep-groups>
```

adriangb commented:
> special support for using path dependencies in this way (with a superproject that delegates to subprojects) is not desirable because it is ultimately incomplete unless we tear down the metadata barrier, and all the designs along that line have been rejected to date

I haven’t heard any objection to my design. If you can think of any issues let's discuss in #6850.

> I think it would have to be comprehensive and not piecemeal

Yes this is a small isolated piece, but that is by design. Part of the attractive of this design is that it can be implemented in small incremental steps, most of which except maybe this one are already completed and we’re beneficial to Poetry regardless of the acceptance of the proposed monorepo design. So I wouldn’t really call it “piecemeal”.

> I'm afraid I still don't see how what you want to do is not possible with dependency groups. Could one not do the following (once https://github.com/python-poetry/poetry-core/pull/520 is in, with all your groups set to optional)?

I don’t think things will work properly without this PR but I’m away from a computer to confirm, I can test later or feel free to give it a crack with my example repo.

adriangb commented:
Ok so I had a chance to look at your example and test in my example repo.

The issue with the solution you proposed is that you never install the transitive 3rd party dependencies (which may even be all 3rd party dependencies if there is no "root" pacakge/dependencies).

I updated my example repo to use this branch (https://github.com/adriangb/python-monorepo/commit/0250ddac28d8cad20381948efad6eae3061f99be) and tested that indeed doing `poetry install --only app --no-path` gives me transitive dependencies of `app` (in particular, `app` depends on `lib` and `lib` depends on `pydantic` and indeed `pydantic` gets installed).

So at least as proposed dependency groups are not an alternative to this PR.

adriangb commented:
pre-commit.ci autofix

adriangb commented:
@radoering I renamed `path` to `directory` as per your feedback in #6850

adriangb commented:
I think this will also be useful in CI workflows (so not Docker but similar) since the same principle of transitive deps changing less than source code apply.

taynaud commented:
Maybe this change is more about top level paths dependancy than directory or files (i.e from the project.toml rather than transitive path dependancies) ? In this case, the option could be --no-root-paths or --no-top-level-paths ?

adriangb commented:
It does skip transitive dir dependencies

mwgamble commented:
> The issue with the solution you proposed is that you never install the transitive 3rd party dependencies (which may even be all 3rd party dependencies if there is no "root" pacakge/dependencies), so there's no layer that has "all of the things that are expensive to install but don't change often" which is what we want for good caching.

Can confirm, this behaviour is essential for making this work in a useful manner. I tried the package groups functionality when I was trying to make this work nicely, and it just didn't work properly at all because of the transitive dependencies issue.

radoering commented on src/poetry/console/commands/install.py:
```diff
@@ -30,6 +30,17 @@ class InstallCommand(InstallerCommand):
+            "no-directory",
+            None,
+            (
+                "Do not install any directory path dependencies (ones using `package ="
+                ' { path = "..." }`\'; useful to install dependencies without source'
```
```suggestion
                "Do not install any directory path dependencies; useful to install dependencies without source'
```

I'd leave that away and expect a user who doesn't know what it is to not use this option or consult the documentation.

radoering commented on tests/console/commands/test_install.py:
```diff
@@ -347,3 +347,15 @@ def test_install_path_dependency_does_not_exist(
             tester.execute(options)
+
+
+def test_no_path_is_passed_to_installer(tester: CommandTester, mocker: MockerFixture):
```
```suggestion
def test_no_directory_is_passed_to_installer(tester: CommandTester, mocker: MockerFixture):
```

radoering replied:
Can you move this somewhere near the other "pass to the installer" tests, e.g. after `test_compile_option_is_passed_to_the_installer` and add a negative test via `pytest.mark.parametrize`  (`_skip_directory` is `False` if option is not passed)

adriangb replied:
donzo

radoering commented on tests/console/commands/test_install.py:
```diff
@@ -347,3 +347,15 @@ def test_install_path_dependency_does_not_exist(
+
+def test_no_path_is_passed_to_installer(tester: CommandTester, mocker: MockerFixture):
+    """
+    The --no-root options is passed to the installer.
```
```suggestion
    The --no-directory option is passed to the installer.
```

radoering commented on src/poetry/puzzle/transaction.py:
```diff
@@ -27,7 +27,10 @@ def __init__(
+        self,
+        with_uninstalls: bool = True,
+        synchronize: bool = False,
+        skip_directory: bool = False,
```
```suggestion
        *,
        skip_directory: bool = False,
```

radoering commented on tests/installation/test_installer.py:
```diff
@@ -1283,6 +1283,49 @@ def test_run_installs_with_local_poetry_directory_transitive(
+    locker: Locker,
+    repo: Repository,
+    package: ProjectPackage,
+    tmpdir: Path,
```
```suggestion
```

unused

radoering commented on tests/installation/test_installer.py:
```diff
@@ -1283,6 +1283,49 @@ def test_run_installs_with_local_poetry_directory_transitive(
     assert installer.executor.installations_count == 6
 
 
+def test_run_installs_with_local_poetry_directory_transitive_no_path(
```
Can you integrate this with `pytest.mark.parametrize` in the previous test?

`skip_directory=False` -> `directory_installs`, `installations_count == 6`
`skip_directory=True` -> `not directory_installs`, `installations_count == ...`

adriangb replied:
did it

radoering reviewed (changes requested):
On the whole, the feature makes sense to me for the mentioned use case and the change is quite minimal. So I think we can merge it after some minor changes.

adriangb commented on docs/cli.md:
````diff
@@ -225,6 +225,12 @@ If you want to skip this installation, use the `--no-root` option.
 poetry install --no-root
 ```
 
+Similar to `--no-root` you can use `--no-directory` to skip path dependencies:
````
469d246b

adriangb commented:
@rodering, I addressed all of the feedback and got the branch in sync with the default branch. Apologies if the commits got messy, I'm working from plane wifi.

adriangb commented:
Thank you for pushing this across the finish line Randy!
